### ISSUE TITLE
Performance: Explore possibilities for memoization

### ISSUE THREAD

**Issue by ranocha** [ranocha/BSeries.jl#4]
`substitute` needs to iterate over the partition forests and skeletons of trees. Currently, these are constructed on the fly each time `substitute` is called. Since this is likely the most expensive part, it might make sense to store values that are already computed.

**Comment by ketch** [comment]
There is an older Python library for working with B-series, and the author used memoization in several places:

https://github.com/henriksu/pybs

**Comment by ranocha** [comment]
Did you compare its performance to your BSeries package?

**Comment by ketch** [comment]
No, not yet. I discovered it long after I had written my own package.

**Comment by ketch** [comment]
Another relevant package is Valentin Dallerit's code: https://gitlab.com/v_dallerit/orderconditions
Docs: https://v_dallerit.gitlab.io/orderconditions/

He has the composition and substitution rules implemented based on the Butcher product, which (I think) implicitly uses memoization since all results on a given tree are built up from subtrees.
